### ISSUE TITLE
[bug] signalSlice: unable to call actionSources if effects are defined above actionSources in the state declaration

### ISSUE THREAD

**Issue by eac-nathan-price** [nartc/ngxtension-platform#235]
@joshuamorony Seems to be the same issue as #213 but with effects instead of selectors. If the effects are defined in the state higher than the actionSources then the actionSources don't get added to the state properly.

**Comment by joshuamorony** [comment]
Thanks I'll see if I can work around this for now, but in the medium term we have this TypeScript 5.4 feature to look forward to: https://twitter.com/mattpocockuk/status/1747182504082116979?s=46 this should greatly simplify the typing of this utility (and hopefully make it more bullet proof in regard to the way generics are ordered/supplied)

**Comment by eac-nathan-price** [comment]
@joshuamorony Good to know! In the meantime having the effects defined below actionSources is an acceptable workaround, I mainly just wanted to make sure that the buggy behaviour was documented.

**Comment by eneajaho** [comment]
@eac-nathan-price Feel free to send a PR to document this workaround. That would be really helpful for others 😀

**Comment by eac-nathan-price** [comment]
@eneajaho Sorry I'm not sure how to make a PR since the workaround is how a user's code can call signalSlice rather than a change to the actual ngxtension/signal-slice library. 

This will cause the actionSources to be bugged:

    state = signalSlice({
      initialState: { ... },
      effects: (state) => ({ ... }),
      actionSources: { ... }
    })

This will work properly:

    state = signalSlice({
      initialState: { ... },
      actionSources: { ... },
      effects: (state) => ({ ... })
    })
    
@joshuamorony identified that changing the order of the keys in the object would bypass the issue when I was having basically the same problem with selectors in #213

**Comment by joshuamorony** [comment]
Both `actionEffects` and `effects` will generically infer the types for `TActionSources` so I'm pretty sure it's currently impossible to fix this (without TypeScript 5.4) without losing the ability to also have typing informaiton for `effects` and `actionEffects`:

```ts
	effects?: (
		state: EffectsState<TSignalValue, TActionSources, TSelectors>,
	) => TEffects;
	actionEffects?: (
		state: EffectsState<TSignalValue, TActionSources, TSelectors>,

```

Any of order configurations should generally work **except** that `effects` and `actionEffects` currently have to be supplied **last**.

I'm pushing a note for this in the docs now.
